scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+
+---
+```ts
+// type: Effect<never, never, number>
+const runnable = program.pipe(
+  Effect.provideService(
+    Random,
+    Random.of({ next: () => Effect.succeed(Math.random()) })
+  )
+);
+
+Effect.runSync(runnable);
+// console: random number: 0.8132812328994277
+```
+notes:
+We can do this with the provide service function. It takes the tag, and a value that fulfils the type of the tag. After we provide the function notice how the requirements field has gone back to never, meaning this Effect can now be run.
+
+Now lets apply this concept to our pokemon program
+
+---
+```ts
+type PokemonClient = {
+  getById(
+    id: number
+  ): Effect.Effect<never, FetchError | JSONError | ParseError, Pokemon>;
+};
+const PokemonClient = Context.Tag<PokemonClient>();
+```
+
+notes:
+Well start by defining the type signature of our client and its tag. For now it will just have 1 method 'getById' which will take a number and return an Effect containing either a Pokemon or the possible errors from before.
+
+---
+
+```ts
+// return type: Effect<PokemonClient, never, Pokemon>
+const getPokemon = (id: number) =>
+  pipe(
+    PokemonClient,
+    Effect.flatMap((client) => client.getById(id)),
+    Effect.catchAll(() => Effect.succeed({ name: "default", weight: 0 }))
+  );
+```
+
+notes:
+Our getPokemon function now looks much simpler as the data fetching logic has been abstracted away. We use the client tag and call the getById method, then catch any errors with our default pokemon. The pokemon client type now appears in the requirements field.
+
+---
+```ts
+program.pipe(
+  Effect.provideService(PokemonClient, {
+    getById: (id) =>
+      pipe(
+        Effect.tryPromise({
+          try: () => fetch(`https://pokeapi.co/api/v2/pokemon/${id}`),
+          catch: () => new FetchError(),
+        }),
+        Effect.flatMap((response) =>
+          Effect.tryPromise({
+            try: () => response.json(),
+            catch: () => new JSONError(),
+          })
+        ),
+        Effect.flatMap(parsePokemon)
+      ),
+  }),
+  Effect.runPromise
+);
+```
+
+notes:
+now we can provide the pokemon client implementation before passing the result to runPromise. Here our implementation of the getById methos is identical to the logic of our getPokemon function previously, but now we could change the data source or adding caching and our downstream usage of the client doesn't need to change at all.
+
+---
+
+```ts
+const getPokemon = (id: number) =>
+  Effect.gen(function* (_) {
+    const client = yield* _(PokemonClient);
+    return yield* _(client.getById(id));
+  }).pipe(
+    Effect.catchAll(() => Effect.succeed({ name: "default", weight: 0 }))
+  );
+```
+
+notes:
+With generators we get the client by yielding the tag and return the yielded result of getById, then provide the service the same as we did just now in the pipe version.
+
+---
+
+## Logging
```

Comment:
I'd probably move this close to the first section so that you can use the Effect.log function in place of console.log within the examples